### PR TITLE
TaskSuspender utility

### PR DESCRIPTION
**Pull request by mihails-strasuns**
Based on idea by @daniel-zullo-sociomantic

### CODE REVIEW

File: src/ocean/task/Scheduler.d [hunk 1]
```diff
@@ -660,6 +660,10 @@ final class Scheduler : IScheduler
                 task.kill();
             else
                 this.resumeTask(task);
+
+            // if calling last task resulted in a shutdown
+            if (this.state == State.Shutdown)
+                break;
```
Comment (daniel-zullo) [inline src/ocean/task/Scheduler.d:664]: would it be possible to add the `break` statement right after `task.kill()`? or could `this.resumeTask(task)` change the state of `this.state`?

File: src/ocean/task/util/TaskSuspender.d [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    /***************************************************************************
+
+        Implements resuming as delayed resuming
+
+    ***************************************************************************/
+
+    override public void resume ( )
+    {
+        theScheduler.delayedResume(task);
```
Comment (daniel-zullo) [inline src/ocean/task/util/TaskSuspender.d]: you might want to prepend `this.`  to task

**Review by daniel-zullo — COMMENTED** [review]
I only have a cosmetic comment and a question and LGTM.
I'll test this patch in one application and report back. Thanks a lot!

File: src/ocean/task/Scheduler.d (hunk 1 above)
Comment (mihails-strasuns) [reply src/ocean/task/Scheduler.d:664]: Yep, it is for the latter - tesumed task calls shutdown internally and scheduler gets into shutdown state on the middle of processing suspend queue.

Comment (daniel-zullo) [reply src/ocean/task/Scheduler.d:664]: right, then alternatively we could have

```D
        if (this.state != State.Shutdown)
            this.resumeTask(task);
                   
        // if calling last task resulted in a shutdown
        if (this.state == State.Shutdown)
        {
            task.kill();
            break;
        }
```
But I'm not sure that makes it better. It's up to you...

**Comment by daniel-zullo** [comment]
I've run a couple of tests and went all good.

File: src/ocean/task/Scheduler.d (hunk 1 above)
Comment (mihails-strasuns) [reply src/ocean/task/Scheduler.d:664]: I have realized that original check is actually not needed at all - calling `shutdown` will kill all suspended tasks anyway. Updated to remove it completely.

**Pull request merged**